apps/prairielearn/src/ee/lib/ai-grading.ts

```diff
@@ -109,41 +261,46 @@ export async function aiGrade({
         job.error('Error occurred');
         job.fail('Errors occurred while AI grading, see output for details');
       }
-      let $ = cheerio.load(render_question_results.data.questionHtml, null, false);
+      const $ = cheerio.load(render_question_results.data.questionHtml, null, false);
       $('script').remove();
       const question_prompt = $.html();
 
-      const render_submission_results = await questionModule.render(
-        { question: false, submissions: true, answer: false },
-        variant,
-        question,
-        submission,
-        [submission],
-        question_course,
-        urls,
+      const submission_embedding = await queryRow(
+        sql.select_embedding_for_submission,
```

Comment:
There's a race condition here: if a student makes a new submission in between `getAllSubmissionEmbeddings` and here, this will fail to return an embedding. Consider handling that gracefully so as not to fail the job as a whole.

Reply:
Now we've moved the select instances for manual grading query to before generating all the embeddings. This way, if a student submits something after the start of the job, that submission will just be not graded and won't give an error. The instructor can click grade again if needed. In the future, we can add warning messages at the end of the job if someone submits after the start of the job.

Reply:
Could you more specifically explain what you changed? AFAICT, the race still exists, since the result of `select_last_variant_and_submission` could change between when that's queried in `generateSubmissionEmbeddings` and above in this loop.

Reply:
Oh crap. We forgot the submission can still change after getting the instance questions. 

I don't have the time to write this now, but here's what I'm proposing. After I run `select_last_variant_and_submission`, I call a variation of `generateSubmissionEmbeddings` that only focuses on one submission, and update the embedding of this submission if needed. Would that sound reasonable?

Reply:
More or less, yeah!

Reply:
For this part, I'm adding a returning statement to the insert query to also get this new row so I don't have to run another query to retrieve this embedding information. I've tested it locally and it is able to perform the insert and return nicely.